src/main/drivers/compass/compass_ist8310.c

```diff
@@ -0,0 +1,256 @@
+
+static bool ist8310Init(magDev_t *magDev)
+{
+    extDevice_t *dev = &magDev->dev;
+    uint8_t regTemp;
+
+    busDeviceRegister(dev);
+
+    // soft-Reset
+    bool ack = busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp |= IST8310_CNTRL2_RESET;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(30);
+
+    // Init setting : avg16 / pulse mode
+    ack = ack && busWriteRegister(dev, IST8310_REG_AVERAGE, IST8310_AVG_16);
+    delay(5);
+    ack = ack && busWriteRegister(dev, IST8310_REG_PDCNTL, IST8310_PULSE_DURATION_NORMAL);
+    delay(5);
+
+    // DRDY enable
+    ack = ack && busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp |= IST8310_CNTRL2_DRENA;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(5);
+
+    // DRDY polarity
+    ack = ack && busReadRegisterBuffer(dev, IST8310_REG_CNTRL2, &regTemp, 1);
+    regTemp &= ~IST8310_CNTRL2_DRPOL;
+    ack = ack && busWriteRegister(dev, IST8310_REG_CNTRL2, regTemp);
+    delay(5);
+
+    return ack;
+}
+
+static bool ist8310Read(magDev_t * magDev, int16_t *magData)
+{
+    extDevice_t *dev = &magDev->dev;
+
+    uint8_t buf[6];
+    // uint8_t LSB2FSV = 3; // 3mG - 14 bit
+    static uint8_t status;
+    static enum {
+        STATE_INIT_DATA,
+        STATE_REQUEST_DATA,
+        STATE_READ_STATUS,
+        STATE_WAIT_STATUS,
+        STATE_WAIT_READ,
+    } state = STATE_INIT_DATA;
+
+    switch (state) {
+        default:
+        case STATE_INIT_DATA:
+            magData[X] = 0;
+            magData[Y] = 0;
+            magData[Z] = 0;
+
+            state = STATE_REQUEST_DATA;
+
+            return false;
+        case STATE_REQUEST_DATA:
+            busWriteRegister(dev, IST8310_REG_CNTRL1, IST8310_ODR_SINGLE);
+            delay(6);
+            state = STATE_READ_STATUS;
+
+            return false;
+        case STATE_READ_STATUS:
+            busReadRegisterBufferStart(dev, IST8310_REG_STAT1, &status, sizeof(status));
+            state = STATE_WAIT_STATUS;
+
+            return false;
+        case STATE_WAIT_STATUS:
+            if ((status & IST8310_DRDY_MASK) == 0) {
+                state = STATE_READ_STATUS;
+                return false;
+            }
+
+            busReadRegisterBufferStart(dev, IST8310_REG_DATA, buf, sizeof(buf));
+            state = STATE_WAIT_READ;
+
+            return false;
+        case STATE_WAIT_READ:
+            // Looks like datasheet is incorrect and we need to invert Y axis to conform to right hand rule
+            // magData[X] =  (int16_t)(buf[1] << 8 | buf[0]) * LSB2FSV;
+            // magData[Y] = -(int16_t)(buf[3] << 8 | buf[2]) * LSB2FSV;
+            // magData[Z] =  (int16_t)(buf[5] << 8 | buf[4]) * LSB2FSV;
+
+            magData[X] = (int16_t)(buf[1] << 8 | buf[0]);
+            magData[Y] = (int16_t)(buf[3] << 8 | buf[2]);
+            magData[Z] = (int16_t)(buf[5] << 8 | buf[4]);
+
+            state = STATE_READ_STATUS;
```

Comment:
This must be `STATE_INIT_DATA`?

Reply:
If it goes to `STATE_INIT_DATA` it resets the MAG values to zero, until it's read, so better `STATE_REQUEST_DATA`.